src/Particle/Par_SetParUID.cpp

```diff
@@ -0,0 +1,144 @@
+#include "GAMER.h"
+
+
+
+#ifdef PARTICLE
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Par_SetParUID
+// Description :  Set the particle unique UID
+//
+// Note        :  1. This function should be done before initializaion
+//                2. Invoked by Init_GAMER() and SF_CreateStar()
+//                3. Assuming all the particles is active when initializing
+//                4. The new particle UID should be initialized as -1
+//                5. Currently, the new particles UID are given by the sorted position
+//
+// Paramter    :  init : Initialization stage or not
+//-------------------------------------------------------------------------------------------------------
+void Par_SetParUID( const bool init )
+{
+
+   long NPar_ThisRank = amr->Par->NPar_AcPlusInac;
+   int  NSend[MPI_NRank], SendDisp[MPI_NRank];
+
+   if ( init )
+   {
+      if ( MPI_Rank == 0 )   Aux_Message( stdout, "Par_SetParID (init) ...\n" );
+
+      long NPar_AllRank;
+      MPI_Allreduce( &NPar_ThisRank, &NPar_AllRank, 1, MPI_LONG, MPI_SUM, MPI_COMM_WORLD );
+
+//    assign UID for all ranks
+      long_par *ParPUid_AllRank = new long_par [NPar_AllRank];
+      for (long p=0; p<NPar_AllRank; p++)   ParPUid_AllRank[p] = (long_par)(p+1);
+
+//    calculate displacement
+      int NPar_ThisRank_int = NPar_ThisRank;
+      MPI_Gather( &NPar_ThisRank_int, 1, MPI_INT, NSend, 1, MPI_INT, 0, MPI_COMM_WORLD );
+      if ( MPI_Rank == 0 )
+      {
+         SendDisp[0] = 0;
+         for (int r=1; r<MPI_NRank; r++)   SendDisp[r] = SendDisp[r-1] + NSend[r-1];
+      } // if ( MPI_Rank == 0 )
+
+//    scatter particle UID to all ranks
+      MPI_Scatterv( ParPUid_AllRank, NSend, SendDisp, MPI_GAMER_LONG_PAR,
+                    amr->Par->PUid,  NPar_ThisRank,   MPI_GAMER_LONG_PAR,
+                    0, MPI_COMM_WORLD );
+
+//    update the next UID to all ranks
+      amr->Par->NextUID = (long_par)(NPar_AllRank+1);
+
+      delete [] ParPUid_AllRank;
+
+      if ( MPI_Rank == 0 )   Aux_Message( stdout, "Par_SetParID (init) ... done\n" );
+   }
+   else // if ( init )
+   {
+      long NNewPar_ThisRank = 0L, NNewPar_AllRank;
+
+//    calculate number of new particles
+      long *NewParIDList = new long [NPar_ThisRank];
+      for (long p=0; p<NPar_ThisRank; p++)
+      {
+         if ( amr->Par->PUid[p] != (long_par)-1 )   continue;
```

Comment:
It seems it assumes the new particles are set with `PUid = -1`. This relies on the correct initialization in the star formation routines. Should we add a sanity check about it in this function? Maybe some particles have other invalid PUids (e.g., -2) accidentally.